File: divide-chocolate/divide-chocolate.py
class Solution:
    def maximizeSweetness(self, sweetness: List[int], k: int) -> int:
        num_of_people = k + 1
        left, right = min(sweetness), sum(sweetness) // num_of_people
        
        while left <= right:
            mid = (left + right + 1) // 2
            cur_sweetness, people_with_chocolate = 0, 0
            
            for s in sweetness:
                cur_sweetness += s
                
                if cur_sweetness >= mid:
                    people_with_chocolate += 1
                    cur_sweetness = 0
                
            if people_with_chocolate >= num_of_people:
                left = mid + 1
            else:
                right = mid - 1
        
        return right
        

File: divide-chocolate/divide-chocolate.cpp
class Solution {
public:
    int maximizeSweetness(vector<int>& sweetness, int k) {
        int n = k + 1;
        int left = *min_element(sweetness.begin(), sweetness.end());
        int right = accumulate(sweetness.begin(), sweetness.end(), 0) / n;
        
        while (left < right){
            int mid = (left + right + 1) / 2, currSweetness = 0, pplWithChocolate = 0;
            for (int s : sweetness){
                currSweetness += s;
                if (currSweetness >= mid){
                    ++pplWithChocolate;
                    currSweetness = 0;
                }
            }
            
            if (pplWithChocolate >= n) left = mid;
            else right = mid - 1;
        }
        
        return right;
    }
};